### ISSUE TITLE
Conversation  [ Not an issue ]

### ISSUE THREAD

**Issue by Akianonymus** [icy/bash-coding-style#9]
Read your guide, agree with most of them.

Here are some of my thoughts which differ a bit:

Function naming, why it's bad to use camelCase ?
also, it's [camelCase](https://en.m.wikipedia.org/wiki/Camel_case) , not CamelCase ( in the guide ).

Error Handling:
About set -e, i think it's really helpful, yes, it's does make the job hard, but one step closer to making the script perfect ( obviously, perfectness is farce ). Handling every possible scenarios is not a bad idea i guess.

Last ( not related to guide )
Recently i refactored a existing bash script, the script is indeed complex for which google recommend to use another language, obviously that is a completely sane advice, but i wanted to do something ridiculous as writing it in bash. Nevertheless, as the guide said, don't hate :)
So, can you just review it ? Just point out some things i am missing in the script ( except set -u, will do it soon )

Link: https://github.com/Akianonymus/google-drive-upload/blob/master/upload.sh

Again, i need to think of my choices that i need to get my bash script reviewed, haha.

**Comment by icy** [comment]
I didn't know there is an issue  (no it is a conversion :) here. I got lost in the Github alerts :(

> Read your guide, agree with most of them.

Great to hear . And thank you very much.

> Here are some of my thoughts which differ a bit:
> 
> Function naming, why it's bad to use camelCase ?
> also, it's [camelCase](https://en.m.wikipedia.org/wiki/Camel_case) , not CamelCase ( in the guide ).

To me, the main reason is that using the underscore (`_`) is very helpful when you view your scripts on remote shell/terminal: I can't expect there is a beautiful editor with color support (and how can we have color support in `less/more`?), or whatever. I have been working with a lot of scripts, and  most of the time, they are on remote machines. Adding a little space helps me to find many issues quickly. 

(It's good for me, hopefully it's good for other people isn't it?)

> Error Handling:
> About set -e, i think it's really helpful, yes, it's does make the job hard, but one step closer to making the script perfect ( obviously, perfectness is farce ). Handling every possible scenarios is not a bad idea i guess.

Yes, they are very helpful. In some Ci system (e,g, Github-Action as I recall), they are turned on by default. The problem may occur when you have a little logic in your script, with `if-then-else`,... In the style guide, there are a few examples when using `set -e` is not very good. To me, the most surprised thing is when my script is terminated (because of `set -e`) without any explanation: the script just stops and unless there is a `set -x` it's so hard to know the cause.

tldr;: It's highly recommended to use `set -e` with `set -x`.

> Last ( not related to guide )

> Recently i refactored a existing bash script, the script is indeed complex for which google recommend to use another language, obviously that is a completely sane advice, but i wanted to do something ridiculous as writing it in bash. Nevertheless, as the guide said, don't hate :)

In my team there is a talent engineer who is using golang, python... He and I want to port some scripts to python and golang for the sake of (`add your 1k reasons here`). We finally recognize that, we are not google, we are not Amazon, we don't have enough time to do that. 

And considering the fact that we can write 600 Bash+Golang lines to solve all crazy problems in our `cd` system, and the script can completely beat this amazing tool (https://github.com/fluxcd/flux) written in Golang , I have no good reason to rewrite my tool in golang.

Look  at our project's CLOC (The heartof the project is 291 Bash lines)

```
github.com/AlDanial/cloc v 1.84  T=0.03 s (266.6 files/s, 25522.2 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Bourne Shell                     2             77             49            291
Go                               4             31              4            230
make                             1             10              0             40
Dockerfile                       1              4              1             29
-------------------------------------------------------------------------------
SUM:                             8            122             54            590
```

and flux's CLOC

```
github.com/AlDanial/cloc v 1.84  T=0.49 s (697.9 files/s, 94187.2 lines/s)
--------------------------------------------------------------------------------
Language                      files          blank        comment           code
--------------------------------------------------------------------------------
Go                              236           4233           2630          28780
Markdown                         37           1669              0           5821
YAML                             48             88            246           1724
Bourne Again Shell               11             82             62            376
make                              1             39              4            152
JSON                              3              0              0             81
SVG                               3              0              3             73
Bourne Shell                      2             10             16             44
CSS                               1              4              0             20
--------------------------------------------------------------------------------
SUM:                            342           6125           2961          37071
--------------------------------------------------------------------------------

```
> So, can you just review it ? Just point out some things i am missing in the script ( except set -u, will do it soon )
> Link: https://github.com/Akianonymus/google-drive-upload/blob/master/upload.sh

Thank you for your asking, I can't promise anything, but I can spend a little time at it. 

> Again, i need to think of my choices that i need to get my bash script reviewed, haha.

I think, just go with what you love and what you feel comfortable, and the problem you want to solve. If your team wants to use x,y,z... then it's a team problem. You have to live with that. If there is 5/5 chance, you can just pick up Bash and show the team how cool your script it is. We can't keep talking about the bad or good things without writing anything. Let your code do its beautiful talk ;)

Nice weekend.

**Comment by icy** [comment]
(Just fixed a few typo errors in the comment; I always have problem with my English)

**Comment by Akianonymus** [comment]
> I didn't know there is an issue (no it is a conversion :) here. I got lost in the Github alerts :(

Aah, my bad, should have named the issue something less confusing.

> 
> > Here are some of my thoughts which differ a bit:
> > Function naming, why it's bad to use camelCase ?
> > also, it's [camelCase](https://en.m.wikipedia.org/wiki/Camel_case) , not CamelCase ( in the guide ).
> 
> To me, the main reason is that using the underscore (`_`) is very helpful when you view your scripts on remote shell/terminal: I can't expect there is a beautiful editor with color support (and how can we have color support in `less/more`?), or whatever. I have been working with a lot of scripts, and most of the time, they are on remote machines. Adding a little space helps me to find many issues quickly.
> 
> (It's good for me, hopefully it's good for other people isn't it?)

Well, you totally make sense, now I can't see my fucntion names properly, haha.

Jokes aside, i completely agree, it is indeed a good convention.


> > Error Handling:
> > About set -e, i think it's really helpful, yes, it's does make the job hard, but one step closer to making the script perfect ( obviously, perfectness is farce ). Handling every possible scenarios is not a bad idea i guess.
> 
> Yes, they are very helpful. In some Ci system (e,g, Github-Action as I recall), they are turned on by default. The problem may occur when you have a little logic in your script, with `if-then-else`,... In the style guide, there are a few examples when using `set -e` is not very good. To me, the most surprised thing is when my script is terminated (because of `set -e`) without any explanation: the script just stops and unless there is a `set -x` it's so hard to know the cause.
> 
> tldr;: It's highly recommended to use `set -e` with `set -x`.

Yeah, i do that, i added a debug flag to my script which runs it with set -ex, very helpful for debugging on some ridiculous errors by set -e.


> In my team there is a talent engineer who is using golang, python... He and I want to port some scripts to python and golang for the sake of (`add your 1k reasons here`). We finally recognize that, we are not google, we are not Amazon, we don't have enough time to do that.
> 
> And considering the fact that we can write 600 Bash+Golang lines to solve all crazy problems in our `cd` system, and the script can completely beat this amazing tool (https://github.com/fluxcd/flux) written in Golang , I have no good reason to rewrite my tool in golang.
> 
> Look at our project's CLOC (The heartof the project is 291 Bash lines)
> 
> ```
> github.com/AlDanial/cloc v 1.84  T=0.03 s (266.6 files/s, 25522.2 lines/s)
> -------------------------------------------------------------------------------
> Language                     files          blank        comment           code
> -------------------------------------------------------------------------------
> Bourne Shell                     2             77             49            291
> Go                               4             31              4            230
> make                             1             10              0             40
> Dockerfile                       1              4              1             29
> -------------------------------------------------------------------------------
> SUM:                             8            122             54            590
> ```
> 
> and flux's CLOC
> 
> ```
> github.com/AlDanial/cloc v 1.84  T=0.49 s (697.9 files/s, 94187.2 lines/s)
> --------------------------------------------------------------------------------
> Language                      files          blank        comment           code
> --------------------------------------------------------------------------------
> Go                              236           4233           2630          28780
> Markdown                         37           1669              0           5821
> YAML                             48             88            246           1724
> Bourne Again Shell               11             82             62            376
> make                              1             39              4            152
> JSON                              3              0              0             81
> SVG                               3              0              3             73
> Bourne Shell                      2             10             16             44
> CSS                               1              4              0             20
> --------------------------------------------------------------------------------
> SUM:                            342           6125           2961          37071
> --------------------------------------------------------------------------------
> ```

Damn, that's impressive.

I agree with your reason point, i believe if something is written in x, does it's job, isn't ridiculously slow, don't touch it.

Again, i don't hate golang or python, neither i am obsessed with bash or i am qualified enough to make some serious comments, but the thing is, i believe one should do what's reasonable or convenient.
golang is indeed a impressive language and i plan to learn it myself soon, but doesn't mean i have to hate bash.

P.S: That table is beautiful 🙂

> 
> Thank you for your asking, I can't promise anything, but I can spend a little time at it.

Haha, no problem, take you time, it was a ridiculous request anyway :p

Continuing the conversation, in your guide, i saw you don't like arrays ?

What's the reason for that ? 🤔

**Comment by icy** [comment]
> 
> Well, you totally make sense, now I can't see my fucntion names properly, haha.
> 
> Jokes aside, i completely agree, it is indeed a good convention.

I know a few old engineers. If you know some, feel free to ask if they have a better idea. I meant, when you're getting old, your eyes will be getting worse, and sometimes `CamelCase` or `camelCase` is harder to read. or maybe I'm wrong :D 

> > tldr;: It's highly recommended to use `set -e` with `set -x`.
> 
> Yeah, i do that, i added a debug flag to my script which runs it with set -ex, very helpful for debugging on some ridiculous errors by set -e.

It's mentioned `Be careful when shipping set -e script to the world.` Having permanent configuration `set -x` is too verbose isn't it?

> 
> Damn, that's impressive.

Thanks :)

> I agree with your reason point, i believe if something is written in x, does it's job, isn't ridiculously slow, don't touch it.

I would translate that to _if it does its job and you don't understand how it works please don't touch it_ :D 

> Again, i don't hate golang or python, neither i am obsessed with bash or i am qualified enough to make some serious comments, but the thing is, i believe one should do what's reasonable or convenient.

I also love golang and python. 

> golang is indeed a impressive language and i plan to learn it myself soon, but doesn't mean i have to hate bash.

Great. ^^

> P.S: That table is beautiful slightly_smiling_face

It's generated by `cloc` (https://github.com/AlDanial/cloc). This is written Perl. I hate Perl. I hope some python and golang or even rust engineers can help to rewrite them very soon. The world is better without Perl :D 

(oh it's a joke of course :D)

> > Thank you for your asking, I can't promise anything, but I can spend a little time at it.
> 
> Haha, no problem, take you time, it was a ridiculous request anyway :p

Okay, fair enough :P 

> Continuing the conversation, in your guide, i saw you don't like arrays ?
> 
> What's the reason for that ? thinking

Sometimes I need to use Array in Bash, it's fine. Bash also uses a few arrays (`FUNCNAME`, `PIPESTATUS`, ...) I just limit myself :) 

If you are using arrays in Bash, I think it makes sense to have good convention. I'm okay to add them if you have one :P

**Comment by icy** [comment]
ohWhyTheWolrdJustDon'tWriteEverySentenceInCamelFormInstead?ItIsVeryEasyToReadLet'sGoWithThisNewStyle :))

**Comment by Akianonymus** [comment]
> > > tldr;: It's highly recommended to use `set -e` with `set -x`.
> > 
> > 
> > Yeah, i do that, i added a debug flag to my script which runs it with set -ex, very helpful for debugging on some ridiculous errors by set -e.
> 
> It's mentioned `Be careful when shipping set -e script to the world.` Having permanent configuration `set -x` is too verbose isn't it?

Nah, not shipping with set -x permanently, i added a flag ( ` --debug` ) to trigger it.

see https://github.com/Akianonymus/google-drive-upload/blob/50d22c36585366accb2f2e2e3d184c19d3104bc9/upload.sh#L27

> > I agree with your reason point, i believe if something is written in x, does it's job, isn't ridiculously slow, don't touch it.
> 
> I would translate that to _if it does its job and you don't understand how it works please don't touch it_ :D

Well, that's more correct, i am not much of a lawyer :p

> > P.S: That table is beautiful slightly_smiling_face
> 
> It's generated by `cloc` (https://github.com/AlDanial/cloc). This is written Perl. I hate Perl. I hope some python and golang or even rust engineers can help to rewrite them very soon. The world is better without Perl :D
> 
> (oh it's a joke of course :D)

Nice tool indeed.

Maybe i will try to rewrite it in bash, till i learn a high level language.

( oh it's a joke of course :D or .. )
 
> > Continuing the conversation, in your guide, i saw you don't like arrays ?
> > What's the reason for that ? thinking
> 
> Sometimes I need to use Array in Bash, it's fine. Bash also uses a few arrays (`FUNCNAME`, `PIPESTATUS`, ...) I just limit myself :)
> 
> If you are using arrays in Bash, I think it makes sense to have good convention. I'm okay to add them if you have one :P

Well, I don't use arrays always. Although i have found some great usage, speeding up things:

1.  When you want to count no of items in a variable, rather than using wc -l or grep or some external program, just make a array and "${#array[@]}", simple pure bash ( from bash bible ).

2.  Using associative arrays to remove duplicate items, alternative to uniq, although not much of a use.

But i tweaked it a little bit and have a function in my script which acts a config parser, i maybe not able to explain, see yourself https://github.com/Akianonymus/google-drive-upload/blob/50d22c36585366accb2f2e2e3d184c19d3104bc9/upload.sh#L202

This is really helpful for the repo i am linking, helps me properly manage values in a file, without using any external programs like grep/sed/awk.

Note: I am more of a `i will try to use only bash builtins and features` guy rather than using any external program, but only when sustainable. But most of the time, bash is enough to do the deed.

P.S: Maybe there should be a section on your guide about this ? just a suggestion.

**Comment by Akianonymus** [comment]
> ohWhyTheWolrdJustDon'tWriteEverySentenceInCamelFormInstead?ItIsVeryEasyToReadLet'sGoWithThisNewStyle :))

underscores_are_my_Friend_now :)

rip_Camel_case

**Comment by icy** [comment]
> Well, I don't use arrays always. Although i have found some great usage, speeding up things:
> 
>     1. When you want to count no of items in a variable, rather than using wc -l or 
>     grep or some external program, just make a array and "${#array[@]}", 
>     simple pure bash ( from bash bible ).
> 
>     2. Using associative arrays to remove duplicate items, alternative to uniq, 
>    although not much of a use.
> 
> 
> But i tweaked it a little bit and have a function in my script which acts a config parser, i maybe not able to explain, see yourself https://github.com/Akianonymus/google-drive-upload/blob/50d22c36585366accb2f2e2e3d184c19d3104bc9/upload.sh#L202

You're right. Arrays are very useful. My point is that I don't use array too much to have a regular rule for them. Moreover, Bash already has some arrays and any new rule likely breaks them and/or has an exception. I can't convince myself enough, that's it.

> This is really helpful for the repo i am linking, helps me properly manage values in a file, without using any external programs like grep/sed/awk.

Agreed. And bash4/5 is very powerful.

> Note: I am more of a `i will try to use only bash builtins and features` guy rather than using any external program, but only when sustainable. But most of the time, bash is enough to do the deed.

If that's the case, go with Bourne shell, where you would see a lot more challenges. I made a mistake in my hobby project (`pacapt`), when I rely too much on Bash-4 features, and it's almost impossible for me to port the script to some limited platform now. 

> P.S: Maybe there should be a section on your guide about this ? just a suggestion.

Thank you, but not at this moment ;)

**Comment by icy** [comment]
@Akianonymus Quick question, is your fork the main fork? Let's say if I want to raise an issue, where should I have them? In your fork or the upstream project (https://github.com/labbots/google-drive-upload)?

Thanks a lot.

**Comment by Akianonymus** [comment]
@icy Aah, sorry for late reply, totally missed the notification.

Now, about doing it in posic, i wanted to, but i went ahead and assumed that people are running bash, because let's be rational for a while.

Most of the Operating System these days ship with bash, if not can be installed fairly easily.

Also, using bash gives a performance boost and extra functionality, that's the sole purpose.

Like using mapfile in arrays, having tons of internal builtins so to avoid less forking of external services, etc.

POSIX is mainly present on every unix based systems because it's the core for booting the os and basic usage.

Hundreds of scripts are triggered during boot and make our system usable. So, a lightweight solution is definitely needed.

With bash, we want a script for our use, with less clutter as possible.

Atlast, yes, it all summarises to convenience.

**Comment by Akianonymus** [comment]
@icy Please do it in the main repository, thank you.

**Comment by icy** [comment]
> @icy Please do it in the main repository, thank you.

Thanks, I have a quick PR to enable shellchecks for your project (https://github.com/labbots/google-drive-upload/pull/54 ). If you think that's fine, you need to reconfigure the project to use the Github-Actions ;)

**Comment by Akianonymus** [comment]
> > @icy Please do it in the main repository, thank you.
> 
> Thanks, I have a quick PR to enable shellchecks for your project ([labbots/google-drive-upload#54](https://github.com/labbots/google-drive-upload/pull/54) ). If you think that's fine, you need to reconfigure the project to use the Github-Actions ;)

First, not my project, haha.

Also, the repo have Integrated codacy which checks for all types of files in the repo including marksown.

Tap on the Codacy badge in the readme.

**Comment by Akianonymus** [comment]
> > > @icy Please do it in the main repository, thank you.
> > 
> > 
> > Thanks, I have a quick PR to enable shellchecks for your project ([labbots/google-drive-upload#54](https://github.com/labbots/google-drive-upload/pull/54) ). If you think that's fine, you need to reconfigure the project to use the Github-Actions ;)
> 
> First, not my project, haha.
> 
> Also, the repo have Integrated codacy which checks for all types of files in the repo including marksown.

Tap on the Codacy badge in the readme.

**Comment by Akianonymus** [comment]
> > > @icy Please do it in the main repository, thank you.
> > 
> > 
> > Thanks, I have a quick PR to enable shellchecks for your project ([labbots/google-drive-upload#54](https://github.com/labbots/google-drive-upload/pull/54) ). If you think that's fine, you need to reconfigure the project to use the Github-Actions ;)
> 
> First, not my project, haha.
> 
> Also, the repo have Integrated codacy which checks for all types of files in the repo including marksown.

Tap on the codacy badge in the readme

**Comment by icy** [comment]
@Akianonymus Oh great, I see Codacy but I didn't see if shellcheck is working there, so just gave it a try ;) Thanks for pointing it out.

**Comment by Akianonymus** [comment]
Closing this as i have my questions answered.
